File: backend/agent/bash_unix.go
//go:build unix

package agent

import (
	"os/exec"
	"syscall"
)

// configureProcGroup puts the command in its own process group and, on context
// cancellation, kills the whole group so child processes spawned by the shell
// (pipelines, subshells) are torn down too — not just the bash parent.
//
// Pdeathsig ensures the group is SIGKILL'd if the Enough process exits without
// a clean Abort (terminal close, kill -9 on parent, etc.) so bash jobs cannot
// outlive the app and poison the next session.
func configureProcGroup(cmd *exec.Cmd) {
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Setpgid:   true,
		Pdeathsig: syscall.SIGKILL,
	}
	cmd.Cancel = func() error {
		if cmd.Process == nil {
			return nil
		}
		// Negative pid targets the process group.
		return syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
	}
}

File: backend/agent/bash_unix.ts
// PORT: backend/agent/bash_unix.go

import { ChildProcess } from "node:child_process";
import process from "node:process";

// configureProcGroup is a no-op on Node.js after process is spawned,
// because process attributes must be passed at spawn time.
export function configureProcGroup(cmd: ChildProcess): void {
  // no-op
}

export function killProcessGroup(cmd: ChildProcess): Error | null {
  if (!cmd || !cmd.pid) {
    return null;
  }
  try {
    // Negative pid targets the process group
    process.kill(-cmd.pid, "SIGKILL");
    return null;
  } catch (err) {
    try {
      cmd.kill("SIGKILL");
    } catch {}
    return err instanceof Error ? err : new Error(String(err));
  }
}

/*
PORT STATUS
source path: backend/agent/bash_unix.go
source lines: 29
draft lines: 27
confidence: high
status: phase_b_compile
*/
